### ISSUE TITLE
[Bug]: Nest cannot resolve dependencies specified in the module (when one uses it with TypeORM?)

### ISSUE THREAD

**Issue by rychkog** [nestjs/nest#220]
Hi @kamilmysliwiec!

I'm currently building a project using Nest and TypeORM. But I've been experiencing a strange issue with a broken DI mechanism. It looks to me that it is somehow connected to the Nest <-> TypeORM wiring
Here is the screenshot of my terminal's output

![image](https://user-images.githubusercontent.com/1942440/32169833-bc0345b6-bd7a-11e7-8a0e-379a3258b03f.png)

Below is the branch of my Nest fork, where this issue is reproducible (based on **examples/05-sql-typeorm**).
https://github.com/nestjs/nest/compare/master...rychkog:typeorm-nest-di-resolver-issue

**Brief description**: when I start creating a module that contains couple repositories as providers and connection for them - DI gets broken and gives me a message that you can see on the screenshot above.

If you'll have any questions regarding the issue - feel free to contact me - I'll try to help as much as I can

**Comment by kamilmysliwiec** [comment]
Hi @rychkog,
Let's try to export entire providers objects, for example:
```
exports: [{
   provide: 'PhotoRepositoryToken2',
   useFactory: (connection: Connection) => connection.getRepository(Photo2),
   inject: ['DbConnectionToken'],
}],
```
To avoid repeating you can create a standalone variable.

**Comment by rychkog** [comment]
@kamilmysliwiec Now it works like that in the example project but doesn't in my actual project... 
At the point where there is an import statement for the main Nest module, it just exists node process

At this point node process exits. I am totally confused ...
```
import { ApplicationModule } from './modules/app.module';
```

**Comment by kamilmysliwiec** [comment]
@rychkog we need more context here

**Comment by rychkog** [comment]
@kamilmysliwiec I understand - trying to come up with a more descriptive example

**Comment by kamilmysliwiec** [comment]
@rychkog awesome 🙂

**Comment by rychkog** [comment]
@kamilmysliwiec Good morning! :)
I've spent whole evening yesterday trying to figure out what is the cause of my issue, so I describe it here in hope, that community might benefit from my pain and foolishness  :)

**Symptoms**: 
1. Nest complains dependency injection is broken and it cannot resolve dependencies of some components or that it cannot access property `provide` of `undefined`.
2. App gets transpiled but when started - exits unexpectedly without any errors, without any little clue in the terminal

*Solutions and observations*
- _for issue#1_
  Your [suggestion](https://github.com/nestjs/nest/issues/220#issuecomment-340427175) about having in `exports` entire provider object helped.
  But occasionally I had an issue that sounded like "cannot access property `provide` of `undefined`". Debugging revealed that dependencies array gets extracted by Reflect from `exports` section occasionally contained undefined as one of the values. Maybe this is connected but your workaround helped and I haven't seen this error since then.

- _for issue#2_
  This one weird and old as the Universe - circular dependency :) 

  I had a `RepositoryModule`, that contained a class like this

  ```typescript
  export class DatabaseObject {
    @PrimaryGeneratedColumn()
    id: number;
     ....
  }
  ```

  This class, in turn, was used by another class

  ```typescript
  export class MySuperClass extends DatabaseObject {
    ...
  }
  ```

  and this class was imported in `RepositoryModule` (just stop here and behold the Cycle:) ) for new repository object creation. :)

  Once I've extracted `DatabaseObject` from the module issue gets resolved

Thanks for your support! :)
